Public API for NJT construction from jagged components

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121518
* #113280
* #113279

This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

jbschlosser committed:
Public API for NJT construction from jagged components

[ghstack-poisoned]

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

YuqingJ commented on torch/nested/__init__.py:
```diff
@@ -302,3 +304,87 @@ def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[in
     return nt
+
+
+def nested_tensor_from_jagged(
```
Will there be methods for me to pass in the  min_seqlen, max_seqlen manually? 
The goal is to avoid calculating it from the offsets(gpu) and syncing it to cpu.

soulitzer replied:
With the new infrastructure, it would be possible have a separate API to associate lengths/offsets with cached quantities. 
Tentatively it would look something like `get_global_nt_state().uf.get_metadata(offsets)["max_seqlen"]  = max_seqlen`

soulitzer replied:
We could probably wrap this something simpler: `nt_set_metadata(offsets, "max_seqlen", max_seqlen)` (name bikesheddable)

cpuhrsch replied:
@soulitzer so that means we'd just calculate this once and then cache it automatically? So then @YuqingJ doesn't have to store it, because we calculate it as needed.

soulitzer replied:
I think that is an option, but if I understand correctly @YuqingJ wants to compute it and store it into the cache ahead of time instead of incurring a sync in the middle of the program (though I'm not sure how much of a concern this given that this cost would be amortized over all places we need to use it later).

davidberard98 replied:
> though I'm not sure how much of a concern this given that this cost would be amortized over all places we need to use it later

wanted to point out that depending on the scenario, this might not necessarily be true; e.g. with PT2 you sometimes have a lot of CPU latency at the end of a compiled section (e.g. doing AccumulateGrad, or creating views) - you could imagine that if the sync is triggered near the end of the compiled section, it might cause GPU idleness at the end of the compiled section when the views run

jbschlosser replied:
based on the above discussion, I'm going to punt on this until @soulitzer's nested int work lands

jbschlosser replied:
based on the above discussion, I'm going to punt on this until @soulitzer's nested int work lands

soulitzer commented on torch/nested/__init__.py:
```diff
@@ -302,3 +304,87 @@ def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[in
+) -> Tensor:
+    r"""
+Constructs a jagged layout nested tensor from the given jagged components. The jagged layout
+consists of a required values buffer with the jagged dimension collapsed into a single dimension.
```
nit: collapsed -> packed? (here and other places) collapsed feels like reduced... but maybe that is just me

jbschlosser replied:
yeah packed is better! will change

soulitzer commented on torch/nested/__init__.py:
```diff
@@ -302,3 +304,87 @@ def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[in
+        else:
+            # TODO: Truly support offsets=None at some point?
+            # For now, just convert lengths -> offsets for kernel convenience
+            offsets = F.pad(lengths.cumsum(0), (1, 0))
```
Yeah, we should eventually have better infrastructure to support initialization with any of lengths/offsets and then doing some lazy converison/caching.

soulitzer reviewed (commented):
Nice! small nit on wording

cpuhrsch commented on torch/nested/__init__.py:
```diff
@@ -302,3 +304,87 @@ def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[in
+    values: Tensor,
+    offsets: Optional[Tensor] = None,
+    lengths: Optional[Tensor] = None,
+    jagged_dim: Optional[int] = None,
```
Some might also want to pass the `size` if this is a reconstruction after a decomposition via `.offsets()` and `.values()`.

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot revert -m "Need to fix BC thing" -c ghfirst

jbschlosser commented:
@pytorchbot revert -m "Need to fix BC thing" -c ghfirst

jbschlosser committed:
Update on "Public API for NJT construction from jagged components"


This PR introduces `torch.nested.nested_tensor_from_jagged(values, offsets=None, lengths=None, jagged_dim=1)` (bikeshedding welcome). This is intended to be the main entrypoint for getting an NJT from the `(values, offsets, lengths)` components. The returned NJT is a view of the `values` component.

Note that `torch.nested.nested_tensor()` / `torch.nested.as_nested_tensor()` already exist for constructing an NJT from a list of tensors.

TODO:
* Some doc formatting; suggestions welcome there
* Tests / examples using `jagged_dim != 1`

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i
